src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+        /// Contains the warnings generated while handling the compiler tasks events.
+        /// </summary>
+        private static readonly IList<Warning> Warnings = new List<Warning>();
+
+        // Private methods.
+
+        /// <summary>
+        /// Creates a hierarchical structure that contains the compilation tasks.
+        /// </summary>
+        private static IList<CompilationTaskNode> BuildCompilationTasksHierarchy()
+        {
+            IList<CompilationTaskNode> compilationTasksForest = new List<CompilationTaskNode>();
+            Queue<CompilationTaskNode> toFindChildrenNodes = new Queue<CompilationTaskNode>();
+
+            // First add the roots (top-level tasks) of all trees to the forest.
+
+            foreach (KeyValuePair<string, CompilationTask> entry in CompilationTasks)
+            {
+                if (entry.Value.ParentName == null)
+                {
+                    CompilationTaskNode node = new CompilationTaskNode(entry.Value);
+                    compilationTasksForest.Add(node);
+                    toFindChildrenNodes.Enqueue(node);
+                }
+            }
+
+            // Iterate through the tasks until all of them have been added to the hierarchy.
+
+            while (toFindChildrenNodes.Count > 0)
+            {
+                CompilationTaskNode parentNode = toFindChildrenNodes.Dequeue();
+                foreach (KeyValuePair<string, CompilationTask> entry in CompilationTasks)
+                {
+                    if (parentNode.Task.Name.Equals(entry.Value.ParentName))
+                    {
+                        CompilationTaskNode childNode = new CompilationTaskNode(entry.Value);
+                        parentNode.Children.Add(childNode.Task.Name, childNode);
+                        toFindChildrenNodes.Enqueue(childNode);
+                    }
+                }
+            }
+
+            return compilationTasksForest;
+        }
+
+        /// <summary>
+        /// Handles a compilation task start event.
+        /// </summary>
+        private static void CompilationEventStartHandler(CompilationLoader.CompilationTaskEventArgs eventArgs)
+        {
+            string key = CompilationTask.GenerateKey(eventArgs.ParentTaskName, eventArgs.TaskName);
+            if (CompilationTasks.ContainsKey(key))
+            {
+                Warnings.Add(new Warning(WarningType.ProcessAlreadyExists, key));
+                return;
+            }
+
+            CompilationTasks.Add(key, new CompilationTask(eventArgs.ParentTaskName, eventArgs.TaskName));
+        }
+
+        /// <summary>
+        /// Handles a compilation task end event.
+        /// </summary>
+        private static void CompilationEventEndHandler(CompilationLoader.CompilationTaskEventArgs eventArgs)
+        {
+            string key = CompilationTask.GenerateKey(eventArgs.ParentTaskName, eventArgs.TaskName);
+            if (!CompilationTasks.TryGetValue(key, out CompilationTask process))
```

Comment:
Consider using the 'var' keyword for 'out' arguments.
```suggestion
            if (!CompilationTasks.TryGetValue(key, out var process))
```